cmd/cronosd/cmd/fix-unlucky-tx.go

```diff
@@ -361,7 +382,96 @@ func (db *tmDB) replayTx(appCreator func() *app.App, height int64, txIndex int,
 	}, nil
 }
 
-func (db *tmDB) patchDB(blockResult *tmstate.ABCIResponses, result *abci.TxResult) error {
+func (db *tmDB) getFilePath(height int64, name string) (string, error) {
+	folder := fmt.Sprintf("patch/%s", name)
+	err := os.MkdirAll(folder, os.ModePerm)
+	return fmt.Sprintf("%s/%d.out", folder, height), err
+}
+
+func (db *tmDB) patchFromFile(height int64) (*abci.TxResult, error) {
+	path, err := db.getFilePath(height, "result")
+	if err != nil {
+		return nil, err
+	}
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	var res abci.TxResult
+	err = res.Unmarshal(data)
+	if err != nil {
+		return nil, err
+	}
+	if err := db.txIndexer.Index(&res); err != nil {
+		return nil, err
+	}
+	path, err = db.getFilePath(height, "blockResult")
+	if err != nil {
+		return nil, err
+	}
+	data, err = ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	var blockRes tmstate.ABCIResponses
+	err = blockRes.Unmarshal(data)
+	if err != nil {
+		return nil, err
+	}
+	blockRes.DeliverTxs[res.Index] = &res.Result
+	if err := db.stateStore.SaveABCIResponses(res.Height, &blockRes); err != nil {
+		return nil, err
+	}
+	return &res, nil
+}
+
+func (db *tmDB) patchDB(blockResult *tmstate.ABCIResponses, result *abci.TxResult, height int64, exportOnly bool) (resultErr error) {
```

Comment:
It's simpler to merge `blockResult` and `result` into one struct, and dump to stdout directly, one line per block.

Reply:
can't merge 2 proto files, wrong when encode

Reply:
https://github.com/cosmos/cosmos-sdk/blob/main/snapshots/stream.go#L40
try sth like this, `protoio.NewDelimitedWriter` and `protoio.NewDelimitedReader`, stream of `proto.Message`:
```
tmstate.ABCIResponses
abci.TxResult
tmstate.ABCIResponses
abci.TxResult
...
```

Reply:
I can try, not sure if it's faster

Reply:
it can be managed in a single stream, no need to write to some fixed filenames.

Reply:
stream seems work, we persist to single file?